### DISCUSSION TITLE
setup not called when testing

### DISCUSSION THREAD

**Discussion by trim21** [fastapi/fastapi#8130] [Questions] [answered]
**Describe the bug**

I'm  using `aioredis` as my redis driver and it's initialized by a async function `aioredis.create_redis_pool`.

So, I create and bind connection pool in `startup`, and in router I get it by `Depends`

```python
@app.on_event('startup')
async def setup():
    app.objects = objects
    app.redis_pool = await aioredis.create_redis_pool()

# depends
async def get_redis(request: Request):
    return request.app.redis_pool

# router
@app.get('/')
async def get_session(redis=Depends(get_redis)):
    pass
```

But when I'm running my test code, it raise a `AttributeError: 'FastAPI' object has no attribute 'redis_pool'`

**To Reproduce**
1. Create a file named `app.py`
2. Create a file named `test_app.py`
3. run test with command `pytest`

app.py:
```python

from fastapi import FastAPI, Depends
from starlette.requests import Request

app = FastAPI()


async def fake_connect_pool():
    async def fake_call(x):
        return x

    return fake_call


@app.on_event('startup')
async def setup():
    app.redis_pool = await fake_connect_pool()


# depends
async def get_redis(request: Request):
    return request.app.redis_pool


# router
@app.get('/')
async def get_session(redis=Depends(get_redis)):
    return await redis(233)


if __name__ == '__main__':
    import uvicorn
    uvicorn.run(app, port=8001)
```

test_app.py
```python
from app import app

from starlette.testclient import TestClient

client = TestClient(app)


def test_index():
    response = client.get("/")
    assert response.status_code == 200
    assert response.text == '233'
```

you will ses

```bash
request = <starlette.requests.Request object at 0x0000025884F3FCF8>

    async def get_redis(request: Request):
>       return request.app.redis_pool
E       AttributeError: 'FastAPI' object has no attribute 'redis_pool'

app.py:21: AttributeError
```

**Expected behavior**
No error should be raised, `startup` handler should be called before test case.

**Environment:**
OS: windows

starlette             0.12.0
fastapi               0.28.0
pytest                4.6.2
python 3.7.2

**Comment by trim21** [comment]
...sorry, didn't read documents carefully.

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue :cake: :slightly_smiling_face:

**Comment by heckad** [comment]
Please, give URL to answer in the documentation.

**Comment by trim21** [comment, marked as answer]
@heckad https://fastapi.tiangolo.com/tutorial/testing/

last section: Testing Events, startup and shutdown

need to use TestClient as context manager

**Comment by hawzie197** [comment]
The section on "Testing Events, startup, and shutdown" looks like it has moved. Found it here: https://fastapi.tiangolo.com/advanced/testing-events/?h=+testing+even

**Comment by keyurm-wmt** [comment]
@hawzie197 @Trim21 I getting the same error after view docs and apply thig.
i do not use the main file app I created a new instance of fastApi and in test case i write a startup event 
can you help me ?

**Comment by trim21** [comment]
> @hawzie197 @Trim21 I getting the same error after view docs and apply thig.
> i do not use the main file app I created a new instance of fastApi and in test case i write a startup event
> can you help me ?

I don't use fastapi anymore, this is the only information I know may help now:

https://www.python-httpx.org/async/#startupshutdown-of-asgi-apps
